os/mm/mm_heap/mm_partition_mgr.c

```diff
@@ -115,24 +117,27 @@ void mm_initialize_ram_partitions(void)
 
 int8_t mm_allocate_ram_partition(uint32_t **start_addr, uint32_t *size)
 {
-	int i;
-
-	for (i = 0; i < CONFIG_NUM_APPS; i++) {
-		if (g_mempartitions[i].status == MM_PART_FREE) {
-			g_mempartitions[i].status = MM_PART_USED;
-			*start_addr = (uint32_t *)g_mempartitions[i].start;
-			*size = g_mempartitions[i].size;
-			mvdbg("Allocated partition Start = 0x%x Size = %u bytes\n", *start_addr, *size);
+	if (*size == 0) {
+		*size = g_default_size;
+	}
 
-			/* struct mm_heap_s will be situated at start of partition
-			and heap will be initialized after this */
-			mm_initialize((struct mm_heap_s *)*start_addr, *start_addr + sizeof(struct mm_heap_s), *size - sizeof(struct mm_heap_s));
+#ifdef CONFIG_DEBUG_MM_HEAPINFO
+	ARCH_GET_RET_ADDRESS
+	*start_addr = mm_malloc(&g_pheap, *size + sizeof(struct mm_heap_s), retaddr);
```

Comment:
I worried about memory fragmentation.
For example, let's assume like below. 
If available ram size(g_pheap size) is 100.
App1 required ram size is 40.
App2 required ram size is 50.
In this case, they will be allocated in the ram like below.
+------------------------------------------+
| ... | App1 (40) | App2 (50) | remain (10)  |
+------------------------------------------+
If user updates the App1 with ram size 41, we cannot update at all..

So how about allocating App2 from the end of the ram?
(It is also not a good solution for more than 2 apps.. anyway...)
Please tell me your opinion.

Reply:
You are right. Fragmentation will be a problem. 
But I think this is the best approach to handle completely dynamic sized and dynamic number of partitions. 
If we are not able to allocate a partition due to fragmentation (it might happen during app  update), then we will return ENOMEM error and the binary manager has to restart the system. Then the problem will be solved.